doc/man/lfe_doc.3
.\" Automatically generated by Pandoc 2.11.2
.\"
.TH "lfe_doc" "3" "2016" "" ""
.hy
.SH NAME
.PP
lfe_doc - Lisp Flavoured Erlang (LFE) documentation parser.
.SH SYNOPSIS
.PP
This module provides functions to parse docstrings in LFE module
sources.
.PP
There is no guarantee the internal formats will not change but the
interface functions should stay the same.
.SH EXPORTS
.PP
\f[B]extract_module_docs(Mod, CompilerInfo) -> {ok,ModDocs} |
{error,Errors,[]}\f[R]
.PP
Parse a module\[cq]s docstrings and return module documentation
structure.
.PP
\f[B]save_module_docs(Beam, ModDocs, CompilerInfo) -> Beam\f[R]
.PP
Add the \f[C]\[dq]LDoc\[dq]\f[R] chunk containing \f[C]ModDocs\f[R] to a
module\[cq]s .beam binary.
.PP
\f[B]get_module_docs(Module | Binary) -> {ok,DocChunk} |
{error,Error}\f[R]
.PP
Extract the documentation chunk from a module.
The chunk will be converted to an internal format.
.PP
\f[B]format_docs/1\f[R]
.PP
Take a list of doc strings and generate a list of indented doc lines.
Each doc string is indented separately.
.PP
\f[B]format_error(ErrorDecriptor) -> Chars\f[R]
.PP
Given an \f[C]ErrorDescriptor\f[R], return a deep list of characters
which describe the error.
This function is usually called implicitly when an \f[C]ErrorInfo\f[R]
structure is processed.
See \f[B]lfe_comp(3)\f[R].
.PP
N.B.
Currently, \f[C]format_error/1\f[R] always returns
\f[C]\[dq]doc error\[dq]\f[R].
.SS MODULE DOC ACCESSORS
.PP
\f[B]module_doc(DocChunk) -> [DocString]\f[R]
.PP
\f[B]mf_docs(DocChunk) -> [MacFuncDoc]\f[R]
.PP
\f[B]mf_doc_type(MacFuncDoc) -> function | macro\f[R]
.PP
\f[B]function_docs(DocChunk) -> [FuncDoc]\f[R]
.PP
\f[B]macro_docs(DocChunk) -> [MacDoc]\f[R]
.PP
Extract fields from the module documentation chunk.
.SS FUNCTION DOC ACCESSORS
.PP
\f[B]function_name(FuncDoc) -> Name\f[R]
.PP
\f[B]function_arity(FuncDoc) -> Arity\f[R]
.PP
\f[B]function_line(FuncDoc) -> Line\f[R]
.PP
\f[B]function_patterns(FuncDoc) -> [Pattern]\f[R]
.PP
\f[B]function_doc(FuncDoc) -> [DocString]\f[R]
.PP
Extract fields from a function documentation structure.
.SS MACRO DOC ACCESSORS
.PP
\f[B]macro_name(MacDoc) -> Name\f[R]
.PP
\f[B]macro_line(MacDoc) -> Line\f[R]
.PP
\f[B]macro_patterns(MacDoc) -> [Pattern]\f[R]
.PP
\f[B]macro_doc(MacDoc) -> [DocString]\f[R]
.PP
Extract fields from a macro documentation structure.
.SH SEE ALSO
.PP
\f[B]lfe_comp(3)\f[R], \f[B]lfe_macro(3)\f[R]
.SH AUTHORS
Eric Bailey.


src/ChangeLog
2013-02-12  Robert Virding  <rv@renat.local>

    * <All files>: Changed Copyright notice and now use Apache License
    V2.0.

2011-08-19  Robert Virding  <rv@renat.local>

    * lfe_macro.erl (pass_expand_expr): Only try to expand lists.

2011-07-09  Robert Virding  <rv@renat.local>

    * lfe_trans.erl (from_bitseg): Fix handling of binaries.

2011-07-02  Robert Virding  <rv@renat.local>

    * lfe_eval.erl (match_binary): Fix handling of binary and pattern
    vars.

    * lfe_lint.erl (pat_binary): Fix handling of binary and pattern
    vars.

2011-06-29  Robert Virding  <rv@renat.local>

    * lfe_lint.erl, lfe_codegen.erl, lfe_eval.erl, lfe_trans.erl: Now
    use lfe_bits to parse binaries.

2011-06-26  Robert Virding  <rv@renat.local>

    * lfe_macro.erl (exp_predef): Add match-spec and qlc macros.

2011-06-15  Robert Virding  <rv@renat.local>

    * lfe_shell.erl (macroexpand, macroexpand-1, macroexpand-all): No
    longer shell functions.

    * lfe_lib.erl (del_vbinding): New function.

2011-06-14  Robert Virding  <rv@renat.local>

    * lfe_io_pretty.erl (indent_type): Add funcall.

2011-06-08  Robert Virding  <rv@renat.local>

    * lfe_macro.erl (exp_predef): Add match-spec macro.

2011-06-05  Robert Virding  <rv@renat.local>

    * lfe_macro.erl (exp_predef): Add simple include-lib macro.

2011-05-30  Robert Virding  <rv@renat.local>

    * lfe_macro.erl (exp_defrecord): Add ets/mnesia match pattern
    macro emp-XXX.

    * lfe_eval.erl, lfe_shell.erl, lfe_macro.erl: Restructure
    lfe_macro.erl and redefine interface.

2011-04-24  Robert Virding  <rv@renat.local>

    * lfe_codegen.erl (comp_gtest): Removed unnecessary special cases,
    optimizer fixes this for us.

2011-04-21  Robert Virding  <rv@renat.local>

    * lfe_lint.erl: Remove unused erroneous lfe_lib imports.

2011-04-19  Robert Virding  <rv@renat.local>

    * lfe_macro.erl (exp_predef): Arithmetic and comparison functions
    take multiple arguments.

2011-04-12  Robert Virding  <rv@renat.local>

    * lfe_comp.erl (erl_comp_opts, do_ok_return): Handle
    warnings_as_errors ourselves in same way as vanilla compiler.

2011-04-08  Robert Virding  <rv@renat.local>

    * lfe_comp.erl (when_opt): Change argument order.

2011-03-16  Robert Virding  <rv@renat.local>

    * lfe_macro.erl (expand_append): Expand ++ in such a way as to
    allow use in patterns.
    (expand_defmacro): Allow (defmacro name arg ...).

2010-12-31  Robert Virding  <rv@renat.local>

    * lfe_io_format.erl (control, newline): Handle ~n correctly.

    * lfe_io_pretty.erl (print1): Prettyprint one element lists.

2010-11-25  Robert Virding  <rv@renat>

    * lfe_lint.erl (pat_bitsegs): Use check_foldl instead.
    (gexpr_bitsegs): Use check_foreach instead.

2010-11-04  Robert Virding  <rv@renat.local>

    * lfe_macro.erl (exp_predef): Extend ? macro to handle optional
    timeout and default value.

2010-11-01  Robert Virding  <rv@renat>

    * lfe_scan.xrl: Check creating atoms/symbols and code cleanup.

2010-10-27  Robert Virding  <rv@renat.local>

    * lfe_shell.erl (prompt): Prompt adds node name when alive.

2010-09-06  Robert Virding  <rv@stanislaw.local>

    * lfe_macro.erl (def_record, def_rec_fields): Record macros follow
    new pattern code.
    (expand_defmacro): Add explicit list constructor for args.
    (c_l_tq): Add explicit cons constructor.

    * lfe_codegen.erl (comp_pat, pat_list): Add code to generate
    pattern code for explicit list constructors.

    * lfe_eval.erl (match, match_list): Add code to match explicit
    list constructors.

    * lfe_lint.erl (check_pat, pat_list, check_alias): Add code to
    test for explicit list constructors.

2010-08-25  Robert Virding  <rv@stanislaw.local>

    * lfe_macro.erl (c_tq, c_l_tq, c_b_tq): Add guards to generators
    in comprehensions.

    * lfe_lint.erl (format_error, check_gbody, check_gexpr)
    (gexpr_bitsegs): Better error checking in guards.

2010-07-24  Robert Virding  <rv@host-90-237-128-188.mobileonline.telia.com>

    * lfe_shell.erl (macroexpand, macroexpand_1): Return evaluated
    argument if no expansion.

2010-07-12  Robert Virding  <rv@host-90-233-145-108.mobileonline.telia.com>

    * lfe_scan.xrl: Have separate cases for integer, float and symbol
    atoms, catch errors better.

2010-06-07  Robert Virding  <rv@stanislaw.local>

    * lfe_lint.erl (expr_bitseg, gexpr_bitseg, pat_bitseg): Allow
    strings (integer lists).

    * lfe_codegen.erl (parse_bitsegs): Allow strings (integer lists).

    * lfe_eval.erl (parse_bitsegs): Allow strings (integer lists).

    * lfe_eval.lfe (eval-gbinary): Evaluate binaries in guards.

    * lfe_eval.erl (eval_gbinary): Evaluate binaries in guards.

2010-06-03  Robert Virding  <rv@stanislaw.local>

    * All files: Update copyright to 2010.

2010-04-30  Robert Virding  <rv@stanislaw.local>

    * lfe_eval.lfe (apply-lambda): New name and more logical argument
    order for eval-lambda/4.
    (apply-match-clauses): New name and more logical argument
    order for eval-match-clauses/3.

2010-02-24  Robert Virding  <rv@stanislaw.local>

    * lfe_eval.erl (apply_lambda): New name and more logical argument
    order for eval_lambda/4.
    (apply_match_clauses): New name and more logical argument
    order for eval_match_clauses/3.

    * lfe_macro.erl (expand_clause, expand_ml_clause, exp_predef)
    (c_tq): Guards are now a sequence of tests (when t ...).

2010-02-20  Robert Virding  <rv@stanislaw.local>

    * lfe_io.erl (scan_and_parse): Use new re-entrant parser.

    * lfe_parse.erl: Write to become re-entrant using LL(1) grammar.

2010-02-16  Robert Virding  <rv@stanislaw.local>

    * lfe_eval.lfe (eval-guard, match-when, eval-let): Guards are now
    a sequence of tests (when t ...).

    * lfe_eval.erl (eval_guard, match_when, eval_let): Guards are now
    a sequence of tests (when t ...).

    * lfe_codegen.erl (comp_guard, comp_gtest, comp_clause_body)

src/lfe_doc.erl
%% Copyright (c) 2016-2017 Eric Bailey
%%
%% Licensed under the Apache License, Version 2.0 (the "License");
%% you may not use this file except in compliance with the License.
%% You may obtain a copy of the License at
%%
%%     http://www.apache.org/licenses/LICENSE-2.0
%%
%% Unless required by applicable law or agreed to in writing, software
%% distributed under the License is distributed on an "AS IS" BASIS,
%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
%% See the License for the specific language governing permissions and
%% limitations under the License.

%% File    : lfe_doc.erl
%% Author  : Eric Bailey
%% Purpose : Lisp Flavoured Erlang (LFE) documentation parser.

%% There is no guarantee the internal formats will not change
%% but the interface functions should stay the same.

-module(lfe_doc).

-export([format_error/1]).

-export([make_doc_info/2,make_chunk/2]).
%%-export([extract_module_docs/1,extract_module_docs/2,save_module_docs/3]).

%% Access functions for documentation in modules.
-export([get_module_docs/1,module_doc/1,function_docs/1,macro_docs/1,
         function_docs/3,macro_docs/2,
         function_name/1,function_arity/1,function_line/1,
         function_patterns/1, function_doc/1,
         macro_name/1,macro_line/1,macro_patterns/1,macro_doc/1]).

-export([format_docs/1]).

-import(lists, [map/2,member/2,filter/2,foldl/3,foldr/3,reverse/1]).

-include("lfe_comp.hrl").
-include("lfe_doc.hrl").

-ifdef(EUNIT).
-export([collect_docs/2,pprint/2]).             %Used by prop_lfe_doc

-include_lib("eunit/include/eunit.hrl").

-define(QC_OPTS, [{on_output,fun pprint/2},{numtests,1000},{max_size,10}]).
-define(QC(T,P), {timeout,30,{T,?_assert(proper:quickcheck(P, ?QC_OPTS))}}).
-endif.

-record(st, {moddoc=[],                         %Module docs
             fs=[],                             %Function info
             ms=[],                             %Macro info
             expf=[],                           %Exported functions and macros
             expm=[],
             errors=[]
            }).

%% Errors
format_error(Error) -> io_lib:format("doc error: ~p", [Error]).

%% make_chunk(Defs, Opts) -> {ok,Chunk}.
%% make_doc_info(Defs, Opts) -> {ok,Chunk}.

-spec make_chunk(Defs, Opts) -> {ok,Chunk} when
      Defs :: [{Form,Line}],
      Form :: [_],
      Line :: non_neg_integer(),
      Opts :: [_],
      Chunk :: {string(),binary()}.

-spec make_doc_info(Defs, Opts) -> {ok,DocInfo} | {error,Errors,[]} when
      Defs :: [{Form,Line}],
      Form :: [_],
      Line :: non_neg_integer(),
      Opts :: [_],
      DocInfo :: lfe_docs_v1(),
      Errors :: nonempty_list({error,Line,any()}).

make_chunk(Defs, Opts) ->
    {ok,DocInfo} = make_doc_info(Defs, Opts),
    Chunk = {"LDoc",erlang:term_to_binary(DocInfo, [compressed])},
    {ok,Chunk}.

make_doc_info(Defs, Opts) ->
    St0 = do_forms(Defs, #st{}),
    St1 = trim_info(St0),
    St2 = format_info(St1),
    #st{moddoc=Moddoc,fs=Fdocs,ms=Mdocs} = St2,
    ?DEBUG("#doc: ~p\n", [{Moddoc,Fdocs,Mdocs}], Opts),
    ?IF(St2#st.errors =:= [],
        {ok,#lfe_docs_v1{moduledoc=Moddoc,fdocs=Fdocs,mdocs=Mdocs}},
        {error,St2#st.errors,[]}).

%% trim_info(State) -> State.
%%  Remove all non-exported functions and macros from the state.

trim_info(#st{fs=Fs0,ms=Ms0,expf=Expf,expm=Expm}=St) ->
    Fs1 = filter(fun (#fdoc{name=N,arity=Ar}) ->
                         is_exported(N, Ar, Expf)
                 end, Fs0),
    Ms1 = filter(fun (#mdoc{name=N}) ->
                         ordsets:is_element(N, Expm)
                 end, Ms0),
    St#st{fs=Fs1,ms=Ms1}.

is_exported(N, Ar, Expf) ->
    Expf =:= all orelse ordsets:is_element([N,Ar], Expf).

%% format_info(State) -> State.
%%  Format all the module, function and macro doc strings in the state.

format_info(#st{moddoc=Mdoc0,fs=Fs0,ms=Ms0}=St) ->
    Mdoc1 = format_docs(Mdoc0),
    Fs1 = map(fun (#fdoc{doc=Ds}=Fd) -> Fd#fdoc{doc=format_docs(Ds)} end, Fs0),
    Ms1 = map(fun (#mdoc{doc=Ds}=Md) -> Md#mdoc{doc=format_docs(Ds)} end, Ms0),
    St#st{moddoc=Mdoc1,fs=Fs1,ms=Ms1}.

%% do_forms(Forms, State) -> State.

do_forms(Forms, St) -> foldl(fun do_form/2, St, Forms).

do_form({['define-module',_,Meta,Atts],_}, St) ->
    define_module(Meta, Atts, St);
do_form({['extend-module',Meta,Atts],_}, St) ->
    define_module(Meta, Atts, St);
do_form({['define-function',Name,Meta,Def],Line}, St) ->
    define_function(Name, Meta, Def, Line, St);
do_form({['define-macro',Name,Meta,Def],Line}, St) ->
    define_macro(Name, Meta, Def, Line, St);
do_form(_, St) -> St.                           %Ignore other forms

define_module(Meta, Atts, #st{moddoc=Md0,expf=Expf0,expm=Expm0}=St) ->
    Md1 = collect_docs(Meta, Md0),
    Md2 = collect_docs(Atts, Md1),
    Expf1 = collect_fexports(Atts, Expf0),
    Expm1 = collect_mexports(Atts, Expm0),
    St#st{moddoc=Md2,expf=Expf1,expm=Expm1}.

collect_fexports(Atts, Expf) ->
    Fun = fun ([export|Es], Efs) -> collect_fexport(Es, Efs);
              (_, Efs) -> Efs
          end,
    foldl(Fun, Expf, Atts).

collect_fexport(_, all) -> all;
collect_fexport([all], _) -> all;
collect_fexport(Es, Efs) ->
    foldl(fun ordsets:add_element/2, Efs, Es).

collect_mexports(Atts, Expm) ->
    Fun = fun (['export-macro'|Es], Em) ->
                  foldl(fun ordsets:add_element/2, Em, Es);
              (_, Em) -> Em
          end,
    foldl(Fun, Expm, Atts).

define_function(Name, Meta, Def, Line, #st{fs=Fs}=St) ->
    {Arity,Pats} = get_function_patterns(Def),
    ?IF(exclude_function(Name, Arity, Meta),
        St,
        begin
            Ds = collect_docs(Meta, []),
            Fd = #fdoc{name=Name,arity=Arity,patterns=Pats,doc=Ds,line=Line},
            St#st{fs=[Fd|Fs]}
        end).

define_macro(Name, Meta, Def, Line, #st{ms=Ms}=St) ->
    ?IF(exclude_macro(Name, Meta),
        St,
        begin
            Ds = collect_docs(Meta, []),
            Pats = get_macro_patterns(Def),
            Md = #mdoc{name=Name,patterns=Pats,doc=Ds,line=Line},
            St#st{ms=[Md|Ms]}
        end).

%% func_arity(['lambda',As|_]) -> length(As);
%% func_arity(['match-lambda',[Pats|_]|_]) -> length(Pats).

collect_docs(As, Mdoc) ->
    %% Collect all the docs in all the doc metas/attributes.
    Afun = fun ([doc|Docs], Md) ->
                   Dfun = fun (D, M) -> M ++ [string_to_binary(D)] end,
                   foldl(Dfun, Md, Docs);
               (_, Md) -> Md
           end,
    foldl(Afun, Mdoc, As).

%% exclude_function(Name, Arity, Meta) -> boolean().
%% exclude_macro(Name, Meta) -> boolean().
%%  Return true if a function should be excluded from the docs chunk.

-spec exclude_function(Name, Arity, Meta) -> boolean() when
      Name  :: atom(),
      Arity :: non_neg_integer(),
      Meta  :: list().
-spec exclude_macro(Name, Meta) -> boolean() when
      Name  :: atom(),
      Meta  :: list().

exclude_function('LFE-EXPAND-EXPORTED-MACRO', 3, _)  -> true;
exclude_function(_, _, _) -> false.

exclude_macro('MODULE', _) -> true;
exclude_macro('FILE', _)   -> true;
exclude_macro('LINE', _)   -> true;
exclude_macro(_, _)        -> false.

%% get_function_patterns(LambdaForm) -> {Arity,Patterns}.
%% get_macro_patterns(LambdaForm)    -> Patterns.
%%  Given a {match-,}lambda form, attempt to return its patterns (or
%%  arglist).  N.B. A guard is appended to its pattern and Patterns is
%%  a list of lists.  A macro definition must have two args, the pattern
%%  and the environment.

-spec get_function_patterns(LambdaForm) -> {Arity,Patterns} when
      LambdaForm :: nonempty_list(),
      Arity      :: non_neg_integer(),
      Patterns   :: nonempty_list({pattern(),guard()}).
-spec get_macro_patterns(LambdaForm) -> Patterns when
      LambdaForm :: nonempty_list(),
      Patterns   :: nonempty_list({pattern(),guard()}).

get_function_patterns([lambda,Args|_]) -> {length(Args),[{Args,[]}]};
get_function_patterns(['match-lambda',[Pat|_]=Cl|Cls]) ->
    {length(Pat),do_function_patterns([Cl|Cls], [])}.

do_function_patterns([[Pat,['when'|Guard]|_]|Cls], Acc) ->
    do_function_patterns(Cls, [{Pat,Guard}|Acc]);
do_function_patterns([[Pat|_]|Cls], Acc) ->
    do_function_patterns(Cls, [{Pat,[]}|Acc]);
do_function_patterns([], Acc) -> reverse(Acc).

get_macro_patterns([lambda,[Args,_Env]|_]) -> [Args];
get_macro_patterns(['match-lambda'|Cls])   -> do_macro_patterns(Cls, []).

do_macro_patterns([[[Pat,_Env],['when'|Guard]|_]|Cls], Acc) ->
    do_macro_patterns(Cls, [{Pat,Guard}|Acc]);
do_macro_patterns([[[Pat,_Env]|_]|Cls], Acc) ->
    do_macro_patterns(Cls, [{Pat,[]}|Acc]);
do_macro_patterns([], Acc) -> reverse(Acc).

string_to_binary(Str) when is_list(Str) ->
    unicode:characters_to_binary(Str, utf8, utf8);
string_to_binary(Bin) -> Bin.

%% format_docs([DocString]) -> [DocLine].
%%  Take a list of doc strings and generate a list of indented doc
%%  lines. Each doc string is indented separately. Should it be so?

format_docs(Ds) -> lists:flatmap(fun format_doc/1, Ds).

format_doc(D) ->
    %% Split the string into separate lines, also trims trailing blanks.
    Ls = re:split(D, <<"[ \t]*\n">>, [trim]),
    format_doc_lines(Ls).

format_doc_lines([<<>>|Ls0]) ->                 %First line empty
    case skip_empty_lines(Ls0) of               %Skip lines until text
        {_,[L|_]=Ls1} ->
            C = count_spaces(L),                %Use indentation of this line
            format_doc_lines(Ls1, C);
        {_,[]} -> []
    end;
format_doc_lines([L1|Ls0]) ->                   %First line not empty
    case skip_empty_lines(Ls0) of
        {Els,[L|_]=Ls1} ->
            C = count_spaces(L),                %Use indentation of this line
            %% Include first line as is.
            [L1] ++ Els ++ format_doc_lines(Ls1, C);
        {Els,[]} -> [L1|Els]
    end;
format_doc_lines([]) -> [].

format_doc_lines(Ls, C) -> lists:map(fun (L) -> skip_spaces(L, C) end, Ls).

count_spaces(L) ->
    {match,[{_,C}]} = re:run(L, <<"^ *">>, []),
    C.

skip_spaces(<<$\s,L/binary>>, C) when C > 0 ->
    skip_spaces(L, C-1);
skip_spaces(L, _) -> L.                         %C =:= 0 or no space

skip_empty_lines(Ls) -> lists:splitwith(fun (L) -> L =:= <<>> end, Ls).

%% Access functions for the module doc chunk.

%% get_module_docs(Module | Binary) -> {ok,Chunk} | {error,What}.

get_module_docs(Mod) when is_atom(Mod) ->
    case code:get_object_code(Mod) of
        {Mod,Bin,_} ->
            get_module_chunk(Bin);
        error -> {error,module}                 %Could not find the module
    end;
get_module_docs(Bin) when is_binary(Bin) ->
    get_module_chunk(Bin).

get_module_chunk(Bin) ->
    case beam_lib:chunks(Bin, ["LDoc"], []) of
        {ok,{_,[{"LDoc",Chunk}]}} ->
            {ok,binary_to_term(Chunk)};
        _ -> {error,docs}                       %Could not find the docs chunk
    end.

%% module_doc(Chunk) -> [binary()].
%% mf_docs(Chunk) -> [MacFuncDoc].
%% mf_doc_type(MacFuncDoc) -> function | macro.
%% function_docs(Chunk) -> [FunctionDoc].
%% macro_docs(Chunk) -> [MacroDoc].

module_doc(#lfe_docs_v1{moduledoc=Moddoc}) -> Moddoc.

function_docs(#lfe_docs_v1{fdocs=Fdocs}) -> Fdocs.
macro_docs(#lfe_docs_v1{mdocs=Mdocs}) -> Mdocs.

function_docs(Name, Ar, #lfe_docs_v1{fdocs=Fdocs}) ->
    Fun = fun (#fdoc{name=N,arity=A}) -> (N =/= Name) or (A =/= Ar) end,
    case lists:dropwhile(Fun, Fdocs) of
        [Fd|_] -> {ok,Fd};
        [] -> error
    end.

macro_docs(Name, #lfe_docs_v1{mdocs=Mdocs}) ->
    case lists:dropwhile(fun (#mdoc{name=N}) -> N =/= Name end, Mdocs) of
        [Md|_] -> {ok,Md};
        [] -> error
    end.

%% function_name(FunctionDoc) -> Name.
%% function_arity(FunctionDoc) -> Arity.
%% function_line(FunctionDoc) -> LineNo.
%% function_patterns(FunctionDoc) -> [Pattern].
%% function_doc(FunctionDoc) -> [DocString].
%%  Extract fields from a function doc structure.

function_name(#fdoc{name=Name}) -> Name.
function_arity(#fdoc{arity=Ar}) -> Ar.
function_line(#fdoc{line=Line}) -> Line.
function_patterns(#fdoc{patterns=Ps}) -> Ps.
function_doc(#fdoc{doc=Ds}) -> Ds.

%% macro_name(MacroDoc) -> Name.
%% macro_line(MacroDoc) -> LineNo.
%% macro_patterns(MacroDoc) -> [Pattern].
%% macro_doc(MacroDoc) -> [DocString].
%%  Extract fields from a macr doc structure.

macro_name(#mdoc{name=Name}) -> Name.
macro_line(#mdoc{line=Line}) -> Line.
macro_patterns(#mdoc{name=N,patterns=Ps}) when is_atom(N) -> Ps.
macro_doc(#mdoc{name=N,doc=Ds}) when is_atom(N) -> Ds.

%%%===================================================================
%%% EUnit tests
%%%===================================================================

-ifdef(EUNIT).
parse_test_() ->
    [ ?QC(<<"A lambda definition is parsed correctly.">>,
          prop_lfe_doc:prop_define_lambda())
    , ?QC(<<"A match-lambda definition is parsed correctly.">>,
          prop_lfe_doc:prop_define_match())
    ].

pprint(_Format, [{Def,_Line}]) -> lfe_io:format(user, "~p\n", [Def]);
pprint(Format, Data) -> lfe_io:format(user, Format, Data).
-endif.
